### PR TITLE
Add ability to redirect generator outptut to file

### PR DESCRIPTION
**Pull request by cescgina**
This introduces two cli parameters, log_mode and log_file. The first
gives the user the ability to choose whether to write the ouptut to
stderr, to a file or to both. In case file is selected, the log_file
parameter selects the path to write to.

### CODE REVIEW

File: znoyder/generator.py [hunk 1]
```diff
@@ -155,12 +155,14 @@ def list_existing_osp_projects() -> list:
...
         rmtree(GENERATED_CONFIGS_DIR)
-        LOG.info('Removed the directory: {GENERATED_CONFIGS_DIR}')
+        LOG.info(f'Removed the directory: {GENERATED_CONFIGS_DIR}')
     destination_directory = os.path.dirname(GENERATED_CONFIGS_DIR)
     Path(destination_directory).mkdir(parents=True, exist_ok=True)
 
 
 def main(args) -> None:
+
+    logger.ensure_correct_logger(args)
```
Comment (sdatko) [inline znoyder/generator.py:165]: Can we make it a global parameter that would be set for all subcommands?

File: znoyder/lib/logger.py [hunk 2]
```diff
@@ -54,6 +56,19 @@
 LOG.propagate = False
 
 
+def ensure_correct_logger(args) -> None:
```
Comment (sdatko) [inline znoyder/lib/logger.py:59]: I would rather call it `set_logger_destination` or something like that

**Review by sdatko — CHANGES_REQUESTED** [review]

File: znoyder/generator.py (hunk 1 above)
Comment (cescgina) [reply znoyder/generator.py:165]: Just to clarify: you mean to give the same option to all subcommands, or to make them a general option for znoyder? Both are easily doable, but I feel like the first option would be the most user friendly.

Comment (sdatko) [reply znoyder/generator.py:165]: I was thinking about a general option for Znoyder, as we use the same logger in most of the components already (if not all right now, we will in the future), so IMO that's the better solution.
Why do you think having such option for each component would be more user friendly? I am afraid that would be mostly a redundancy, but maybe I miss something.

Comment (cescgina) [reply znoyder/generator.py:165]: Maybe user friendly is not really the right wording. I actually tried this after your first comment and what I did not like is that the logging options should be specified before the command, that is
`znoyder --log-file output.txt generate --component ...`
which I feel might be confusing and users might forget to put those options before and get errors.

**Review by bregman-arie — CHANGES_REQUESTED** [review]

File: znoyder/generator.py (hunk 1 above)
Comment (sdatko) [reply znoyder/generator.py:165]: Ach, right, I remember something like this, it's some limitation of argparser with subparsers. This is definitely a user-friendly argument. Though I recall I saw some workaround for this, but I am not sure right now, but I think with small refactor of argparsers it should be achievable.

Ok, so I am fine with keeping it how it is and we may propose a dedicated task to research this workaround (or you can do it as part of this task, if you have some sprare working hours :smiley:)

Comment (bregman-arie) [reply znoyder/generator.py:165]: If I may chime in, I think the most user-friendly option is to not let the user think where he should put these - so basically support it before the sub command and after.

**Comment by cescgina** [comment]
I have updated the PR with a new version, with log mode "both" as default, a "znoyder_output.log" as default log file and the parameters can be passed either as a general znoyder option or as an option for the command

File: znoyder/cli.py [hunk 3]
```diff
@@ -22,12 +22,84 @@
...
 
 from znoyder import browser
 from znoyder import downloader
 from znoyder import finder
 from znoyder import generator
 from znoyder import templater
+from znoyder.lib import logger
+
+
+class NoDefaultOverrideSubparserAction(_SubParsersAction):
```
Comment (bregman-arie) [inline znoyder/cli.py:38]: Perhaps I would simply call it SubparserAction (or OverridenSuperarserAction?). But whatever you feel about it.

**Review by bregman-arie — DISMISSED** [review]
Looks good :)

File: znoyder/cli.py (hunk 3 above)
Comment (cescgina) [reply znoyder/cli.py:38]: Yes, that's probably a better name, I tried to be as explicit as possible, but it's too long, I'll change it

**Review by bregman-arie — APPROVED** [review]

**Review by sdatko — APPROVED** [review]
With OverridenSubparserAction this became bigger than I assumed, but overall it looks great. God job!

File: znoyder/cli.py (hunk 3 above)
Comment (sdatko) [reply znoyder/cli.py]: Just to left a note for any visitors: a generally good practice is to put such details in commit message or docstring (which I can see you did, so I am satisfied)

**Pull request merged**